batch/batch/batch.py

```diff
@@ -325,49 +336,66 @@ def job_record_to_dict(record, running_status=None):
-            'attempt_id': attempt_id,
-            'user': record['user'],
-            'state': 'error',
-            'error': traceback.format_exc(),
-            'container_statuses': {k: {} for k in tasks}
-        }
-        await mark_job_complete(app, batch_id, job_id, attempt_id, 'Error', status,
-                                None, None, 'error')
-        return
+        try:
+            body = await job_config(app, record, attempt_id)
+        except Exception:
+            log.exception('while making job config')
+            status = {
+                'worker': None,
+                'batch_id': batch_id,
+                'job_id': job_id,
+                'attempt_id': attempt_id,
+                'user': record['user'],
+                'state': 'error',
+                'error': traceback.format_exc(),
+                'container_statuses': {k: {} for k in tasks}
+            }
+            await mark_job_complete(app, batch_id, job_id, attempt_id, instance.name,
+                                    'Error', status, None, None, 'error')
+            raise
+
+        log.info(f'schedule job {id} on {instance}: made job config')
 
-    log.info(f'schedule job {id} on {instance}: made job config')
+        try:
+            async with aiohttp.ClientSession(
+                    raise_for_status=True, timeout=aiohttp.ClientTimeout(total=60)) as session:
+                url = (f'http://{instance.ip_address}:5000'
+                       f'/api/v1alpha/batches/jobs/create')
+                await session.post(url, json=body)
+                await instance.mark_healthy()
+        except Exception as e:
+            if (isinstance(e, aiohttp.ClientResponseError) and
+                    e.status == 403):  # pylint: disable=no-member
+                await instance.mark_healthy()
+                log.info(f'attempt already exists for job {id} on {instance}, aborting')
+            else:
+                await instance.incr_failed_request_count()
+            raise e
 
-    try:
-        async with aiohttp.ClientSession(
-                raise_for_status=True, timeout=aiohttp.ClientTimeout(total=60)) as session:
-            url = (f'http://{instance.ip_address}:5000'
-                   f'/api/v1alpha/batches/jobs/create')
-            await session.post(url, json=body)
-            await instance.mark_healthy()
-    except Exception:
-        await instance.incr_failed_request_count()
-        raise
+        log.info(f'schedule job {id} on {instance}: called create job')
 
-    log.info(f'schedule job {id} on {instance}: called create job')
+        rv = await db.execute_and_fetchone(
+            '''
+CALL schedule_job(%s, %s, %s, %s);
+''',
+            (batch_id, job_id, attempt_id, instance.name))
+    except:  # pylint: disable=bare-except
+        log.exception(f'error while scheduling job {id} on {instance}')
+        if instance.state == 'active':
+            instance.adjust_free_cores_in_memory(record['cores_mcpu'])
```

Comment:
this feels likely to lead to weird error conditions? Like if someone adds a variable to a log statement which triggers an exception *after* the job is already scheduled on the worker. I guess you're trying to avoid duplicating this if and update a bunch?

Reply:
It would be really nice if we knew exactly the exceptions that might come and we handled all the expected cases.

Reply:
Yep. was trying to avoid duplication.